src/libawkward/Content.cpp

```diff
@@ -1827,22 +1827,32 @@ namespace awkward {
     }
 
     if (parameter_equals("__array__", "\"categorical\"")) {
-      if (dynamic_cast<const IndexedArray32*>(this)  ||
-          dynamic_cast<const IndexedArrayU32*>(this)  ||
-          dynamic_cast<const IndexedArray64*>(this)  ||
-          dynamic_cast<const IndexedOptionArray32*>(this)  ||
-          dynamic_cast<const IndexedOptionArray64*>(this)) {
-        if (!is_unique()) {
-          return (std::string("at ") + path + std::string(" (") + classname()
-                  + std::string("): __array__ = \"categorical\" requires contents "
-                                "to be unique"));
-        }
+      ContentPtr content(nullptr);
+      if (const IndexedArray32* raw = dynamic_cast<const IndexedArray32*>(this)) {
+        content = raw->content();
+      }
+      else if (const IndexedArrayU32* raw = dynamic_cast<const IndexedArrayU32*>(this)) {
+        content = raw->content();
+      }
+      else if (const IndexedArray64* raw = dynamic_cast<const IndexedArray64*>(this)) {
+        content = raw->content();
+      }
+      else if (const IndexedOptionArray32* raw = dynamic_cast<const IndexedOptionArray32*>(this)) {
+        content = raw->content();
+      }
+      else if (const IndexedOptionArray64* raw = dynamic_cast<const IndexedOptionArray64*>(this)) {
+        content = raw->content();
       }
       else {
         return (std::string("at ") + path + std::string(" (") + classname()
                 + std::string("): __array__ = \"categorical\" only allowed for "
                               "IndexedArray and IndexedOptionArray"));
       }
+      if (!content.get()->is_unique()) {
+        return (std::string("at ") + path + std::string(" (") + classname()
+            + std::string("): __array__ = \"categorical\" requires contents "
+                            "to be unique"));
+      }
     }
```

Comment:
This is a proper correction. If an array is labeled `"categorical"`, then it is _only_ allowed to be an IndexedArray or an IndexedOptionArray, and it is the _contents_ that must be unique, not the array itself. It was wrong before, but I also don't see any tests complaining about the change, so it must not have been tested, either. Thanks for fixing that!

Reply:
I see that you have other things to add to this PR; I won't merge it until you let me know that it's ready by un-drafting it. Then I'll check the next set of changes.